「Webセキュリティ」の課題

ham-cap opened the pull request:
「Webセキュリティ」の課題を提出いたします。
ご確認のほどよろしくお願いいたします🙇‍♂️

ham-cap committed:
セットアップ

ham-cap committed:
ブログ検索機能の是弱性を修正

・検索時にSQLインジェクションが可能な状態だったため修正した

ham-cap committed:
XSS対策

ham-cap committed:
他人のブログの編集画面にアクセスできないようにした

ham-cap committed:
他人のブログに対するアクセス制御を修正

・前回のコミットで実装した編集画面へのアクセス制御の方法ではテストが通らなかったため修正した。
・この修正で編集だけでなく削除の制御も行った。

ham-cap committed:
秘密のブログへのアクセス制御を実装

ham-cap committed:
通常ユーザーがアイキャッチを設定できてしまう脆弱性を修正

ham-cap committed:
APIの返却値にユーザーのメールアドレスが含まれている脆弱性を修正

ham-cap committed:
XSS対策箇所を修正

・当初の実装だとgsubで置換した<br>タグが反映されず改行ができなくなっていたため、sanitizeの呼び出し時に<br>のみ許可することとした。

ham-cap committed:
rubocopでオートコレクトを実行

ham-cap committed:
不要なrequireを削除

cafedomancer commented on package-lock.json:
```diff
@@ -0,0 +1,295 @@
+    "js-tokens": {
+      "version": "4.0.0",
+      "resolved": "https://registry.npmjs.org/js-tokens/-/js-tokens-4.0.0.tgz",
+      "integrity": "sha512-RdJUflcE3cUzKiMqQgsCu06FPu9UdIJO0beYbPhHN4k6apgJtifcoCtT9bcxOpYBtpD2kCM6Sbzg4CausW/PKQ=="
+    },
+    "loose-envify": {
+      "version": "1.4.0",
+      "resolved": "https://registry.npmjs.org/loose-envify/-/loose-envify-1.4.0.tgz",
+      "integrity": "sha512-lyuxPGr/Wfhrlem2CL/UcnUc1zcqKAImBDzukY7Y5F/yQiNdko6+fRLevlw1HgMySw7f611UIY408EtxRSoK3Q==",
+      "requires": {
+        "js-tokens": "^3.0.0 || ^4.0.0"
+      }
+    },
+    "react": {
+      "version": "18.2.0",
+      "resolved": "https://registry.npmjs.org/react/-/react-18.2.0.tgz",
+      "integrity": "sha512-/3IjMdb2L9QbBdWiW5e3P2/npwMBaU9mHCSCUzNln0ZCYbcfTsGbTJrU/kGemdH2IWmB2ioZ+zkxtmq6g09fGQ==",
+      "requires": {
+        "loose-envify": "^1.1.0"
+      }
+    },
+    "react-dom": {
+      "version": "18.2.0",
+      "resolved": "https://registry.npmjs.org/react-dom/-/react-dom-18.2.0.tgz",
+      "integrity": "sha512-6IMTriUmvsjHUjNtEDudZfuDQUoWXVxKHhlEGSk81n4YFS+r/Kl99wXiwlVXtPBtJenozv2P+hxDsw9eA7Xo6g==",
+      "requires": {
+        "loose-envify": "^1.1.0",
+        "scheduler": "^0.23.0"
+      }
+    },
+    "scheduler": {
+      "version": "0.23.0",
+      "resolved": "https://registry.npmjs.org/scheduler/-/scheduler-0.23.0.tgz",
+      "integrity": "sha512-CtuThmgHNg7zIZWAXi3AsyIzA3n4xx7aNyjwC2VJldO2LMVDhFK+63xGqq6CsJH4rTAt6/M+N4GhZiDYPx9eUw==",
+      "requires": {
+        "loose-envify": "^1.1.0"
+      }
+    }
+  }
+}
```
間違って npm install を実行していませんか？

ham-cap replied:
セットアップの際に以下のように複数のGemが見つからないというエラーが出たため、いろいろ調べてアプリが動くところまで持って行ったのですが、その過程で`npm install`と`bundle install`を実行したように思います💧
その前段階としてRubyの3.2.2をインストールしているので、その影響だったのだと思いますが、とりあえずアプリを動かすところまでは行けたのであまり気にしていませんでした。申し訳ありません🙇‍♂️
```
/Users/ham-cap/.rbenv/versions/3.2.2/lib/ruby/3.2.0/bundler/definition.rb:524:in `materialize': Could not find rails-7.0.6, sprockets-rails-3.4.2, sqlite3-1.6.3, puma-6.3.0, jsbundling-rails-1.1.2, turbo-rails-1.4.0, stimulus-rails-1.2.1, jbuilder-2.11.5, bootsnap-1.16.0, sassc-rails-2.1.2, debug-1.8.0, erb_lint-0.4.0, rubocop-1.45.1, rubocop-fjord-0.3.0, rubocop-rails-2.20.2, web-console-4.2.0, capybara-3.39.2, selenium-webdriver-4.11.0, devise-4.9.2, devise-i18n-1.11.0, actioncable-7.0.6, actionmailbox-7.0.6, actionmailer-7.0.6, actionpack-7.0.6, actiontext-7.0.6, actionview-7.0.6, activejob-7.0.6, activemodel-7.0.6, activerecord-7.0.6, activestorage-7.0.6, activesupport-7.0.6, railties-7.0.6, sprockets-4.2.0, mini_portile2-2.8.4, nio4r-2.5.9, msgpack-1.7.2, sassc-2.4.0, tilt-2.2.0, irb-1.7.4, reline-0.3.7, better_html-2.0.2, parser-3.2.2.3, rainbow-3.1.1, smart_properties-1.17.0, parallel-1.23.0, regexp_parser-2.8.1, rexml-3.2.6, rubocop-ast-1.29.0, ruby-progressbar-1.13.0, unicode-display_width-2.4.2, rubocop-performance-1.18.0, rack-2.2.8, bindex-0.8.1, addressable-2.8.5, mini_mime-1.1.2, nokogiri-1.15.3, rack-test-2.1.0, xpath-3.2.0, rubyzip-2.3.2, websocket-1.2.9, bcrypt-3.1.19, orm_adapter-0.5.0, responders-3.1.0, warden-1.2.9, websocket-driver-0.7.6, mail-2.8.1, net-imap-0.3.7, rails-dom-testing-2.2.0, rails-html-sanitizer-1.6.0, globalid-1.1.0, builder-3.2.4, erubi-1.12.0, marcel-1.0.2, concurrent-ruby-1.2.2, i18n-1.14.1, minitest-5.19.0, tzinfo-2.0.6, method_source-1.0.0, thor-1.2.2, zeitwerk-2.6.11, ffi-1.15.5, ast-2.4.2, racc-1.7.1, public_suffix-5.0.3, websocket-extensions-0.1.5, loofah-2.21.3, timeout-0.4.0, crass-1.0.6 in locally installed gems (Bundler::GemNotFound)
```

このファイルと`yarn.lock`は今回の修正部分には直接関係ないかと思いますので、過去のコミットからこのファイルを除外してプルリクエストを作り直した方がよいでしょうか？👀

cafedomancer replied:
以下をやれば Pull Request を作り直さなくても対応できると思います！

1. `node_modules` を念のために削除
1. `package-lock.json` を削除、その変更をコミット
1. `yarn.lock` を main ブランチの内容で復元 (`git checkout main yarn.lock` など) 、その変更をコミット (2. と一緒にコミットしても OK)
1. `yarn` を実行して依存関係を再インストール (`bin/setup` でも OK)

ham-cap replied:
対応が遅くなり申し訳ありません🙇‍♂️
もろもろ立て込んでいました。
ご教示いただいた方法でうまくいきました🙌
ありがとうございます🙏
その他、コードの修正は少しお時間いただければ幸いです。

cafedomancer commented on yarn.lock:
```diff
@@ -170,41 +65,41 @@ esbuild@^0.17.19:
+  "version" "0.23.0"
   dependencies:
-    loose-envify "^1.1.0"
+    "loose-envify" "^1.1.0"
```
これは必要のない変更に見えます 👀

cafedomancer commented on app/views/api/liking_users/_user.json.jbuilder:
```diff
@@ -1,2 +1 @@
-json.extract! user, :id, :email, :created_at, :updated_at, :nickname, :premium
-
+json.extract! user, :id, :created_at, :updated_at, :nickname, :premium
```
問題のある部分を削除するのではなくて、必要な情報だけ露出するようにしてみてください。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -50,4 +57,13 @@ def set_blog
+    Blog.written_by(user.id).find_by!(user_id: current_user.id)
+  end
+
+  def ensure_premium
+    @blog.update(random_eyecatch: false) if @blog.random_eyecatch && current_user.premium == false
+  end
```
save や update のあとにこのメソッドを呼ぶ形式だと、データを操作する SQL が 2 回発行されてしまいます。また、トランザクションの観点からは、2 つの SQL はトランザクションになっているわけではないので、先の SQL が成功して後の SQL が失敗すると `:random_eyecatch` が設定できてしまいます。データを修正する方式ではなくて、課金ユーザーでない場合は `:random_eyecatch` のリクエストパラメータを無視するようにするのがよいと思います。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -5,11 +5,16 @@ class BlogsController < ApplicationController
 
-  def show; end
+  def show
+    raise ActiveRecord::RecordNotFound if @blog.secret && current_user.nil?
+    raise ActiveRecord::RecordNotFound if @blog.secret && @blog.user_id != current_user.id
```
条件を満たすようなデータを検索する SQL クエリを考えてみてください。そこから find するようにすれば `ActiveRecord::RecordNotFound` は自分で raise しなくても済むようになるはずです。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -50,4 +57,13 @@ def set_blog
   end
+
+  def ensure_correct_user
+    user = @blog.user
+    Blog.written_by(user.id).find_by!(user_id: current_user.id)
```
`current_user` の関連からブログを検索するようにするとよいと思います。

cafedomancer commented on app/views/blogs/_blog.html.erb:
```diff
@@ -13,7 +13,7 @@
     <% end %>
     <div class="blog-content">
-      <%= format_content(blog) %>
+      <%= sanitize(format_content(blog),tags: %w(br)) %>
```
サニタイズの処理も `format_content` に含めてしまうのはどうでしょうか？

ham-cap committed:
誤ってnpm installを実行したため不要なファイルを削除しyarn.lockを元の状態に復元した

ham-cap committed:
APIの返却値を修正

・ライクしたユーザーがわかればいいため、返却値はユーザーIDのみとした

ham-cap committed:
APIの返却値を再度修正

・ユーザーIDのみを返すように修正したが、必要なのはニックネームだったため再度修正した。

ham-cap committed:
BlogHelper内のformat_contentメソッドを修正

・sanitizeをview側で行なっていたが、format_contentの処理に含めることとした。

ham-cap committed:
BlogsControllerのensure_correct_userを修正

・current_userの関連で取得したブログを検索するように変更した

ham-cap committed:
Blogのshowアクションを修正・Strong Parameterを修正

・ログイン状態によってアクセスできるブログを制限している部分で、権限のないものにアクセスしようとした場合に強制的にエラーをraiseしていたが、条件にマッチするブログを先に取得しておき、そこからブログをfindすることとした。

・ensure_premiumを使用して、不正なパラメータを強制的にupdateする方式の場合、SQLのクエリが2回発行されてしまううえ、何らかの理由でensure_premiumの処理が失敗した場合は不正ができてしまうため、Strong Parameterで不正なリクエストパラメータ自体を無視することとした。

ham-cap commented:
ご指摘いただいた部分を修正いたしましたのでご確認いただけますでしょうか🙇‍♂️

また、[こちら](https://github.com/ham-cap/haunted-blog/pull/2#discussion_r1320572713)のご指摘について一点質問なのですが、エラーを強制的に`raise`させること自体を極力避けた方が良いという理解で大丈夫でしょうか？👀

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -1,15 +1,27 @@
 # frozen_string_literal: true
 
+require 'debug'
```
不要な require が残っています 👀

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -1,15 +1,27 @@
+              Blog.where('secret = ? OR user_id = ?', false, current_user.id)
+            else
+              Blog.published
+            end
```
1 つの SQL にまとめられると思うので、そうしてみてくださいー。

cafedomancer commented on app/views/api/liking_users/_user.json.jbuilder:
```diff
@@ -1,2 +1 @@
-json.extract! user, :id, :email, :created_at, :updated_at, :nickname, :premium
-
+json.extract! user, :nickname
```
ID は使われていると思うので、このデータを使っている箇所を再確認してみてください。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -48,6 +60,11 @@ def set_blog
 
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    params.require(:blog).permit(:title, :content, :secret, current_user.premium ? :random_eyecatch : nil)
```
`nil` を含める必要はないと思うので、そうするようなコードにしてみてください。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -48,6 +60,11 @@ def set_blog
+
+  def ensure_correct_user
+    user = @blog.user
+    current_user.blogs.find_by!(user_id: user.id)
```
これに近いこと (`current_user` の `blogs` から対象のブログを見つける) を `set_blog` で行うようにすると、このメソッドは不要になるんじゃないでしょうか。

ham-cap committed:
不要なrequireを削除

ham-cap committed:
blog_paramsを修正

・random_eyecatchをストロングパラメーターに追加するか否かを決定する処理で、current_user.premiumがfalseの場合にnilを入れる必要はないため三項演算子をやめた。

ham-cap committed:
showメソッドでアクセス権限のあるブログを取得する処理を修正

・current_userの有無によって発行するSQLを切り替えていたが、条件分岐をする必要がないため修正した。

ham-cap committed:
set_blogの処理を変更しensure_correct_userを削除した

・set_blogで全てのブログからparams[:id]に一致するブログを探してからensure_correct_userで適正なユーザーかどうかを判定すると二度手間となるため、set_blogでブログを取得する際に最初からcurrent_userのブログに限定して取得することでユーザーを判断する必要性をなくした。

ham-cap committed:
APIの返却値を修正

・liking_usersで取得したデータがReactコンポーネント内で呼び出される際にidも必要になるため復活させた。

ham-cap commented:
修正いたしましたのでご確認をお願いいたします🙏

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,17 @@
 
-  def show; end
+  def show
+    blogs = Blog.where('secret = ? OR user_id = ?', false, current_user&.id)
```
`secret = false` の部分は既存の scope が使えると思いますー。また、変更後のコードに関する先出しなのですが、belongs_to 関連で検索を行う場合には、ID ではなくオブジェクトそのものを指定することができます。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,17 @@
 
-  def show; end
+  def show
+    blogs = Blog.where('secret = ? OR user_id = ?', false, current_user&.id)
+
+    @blog = blogs.find(params[:id])
```
これは 1 行で書いてしまってよいと思います。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -44,10 +48,11 @@ def destroy
 
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    random_eyecatch = :random_eyecatch if current_user.premium
+    params.require(:blog).permit(:title, :content, :secret, random_eyecatch)
```
`current_user.premium` が `false` のときは `random_eyecatch` に `nil` が格納されてしまうので、https://github.com/ham-cap/haunted-blog/pull/2#discussion_r1333742651 にコメントした問題が解決できていないように思いました 😵 プリントデバッグやデバッガを挟んで `random_eyecatch` に `nil` が格納されていることを確認するのと、問題を解決できる別の方法を使用することを検討してみてください。

ham-cap committed:
showメソッド内の処理を1行にまとめた

ham-cap committed:
blog_paramsを再度修正

・current_user.premiumがfalseの場合にrandom_eyecatchにnilが格納されてしまうため、permitへの引数の渡し方を変更した。

ham-cap committed:
showメソッドを微修正

・current_userのブログを検索する部分でidではなくオブジェクトそのものを渡すこととした。

ham-cap commented:
修正してみました！
何度もすみませんが、よろしくお願いいたします🙇‍♂️

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -44,10 +46,12 @@ def destroy
 
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    arguments = %i[title content secret]
```
`arguments` だとちょっと一般的すぎるので、もう少し具体的な名前をつけたほうがよさそうですー。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -44,10 +46,12 @@ def destroy
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    arguments = %i[title content secret]
+    current_user.premium && arguments << :random_eyecatch
```
`&&` で確かに期待通りの挙動にはなるのですが、読み手としては `&&` はあくまで複合条件を示すときに使われるという期待を持って読むと思います。なので、後置 if で書いたほうが挙動を見通しやすくなるんじゃないかなと思いました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,15 @@
 
-  def show; end
+  def show
+    @blog = Blog.published.or(Blog.where('user_id = ?', current_user)).find(params[:id])
```
文字列でないと指定できない条件以外は、ハッシュ形式で書いたほうが見やすいと思います。

```suggestion
    @blog = Blog.published.or(Blog.where(user: current_user)).find(params[:id])
```

ham-cap committed:
blog_paramsを修正

・変数名がargumentsでは具体性が無さすぎるため変更した。
・random_eyecatchを追加する部分で&&を使用していたが、可読性を考慮して後置ifへ変更した。

ham-cap committed:
showメソッド内のwhere句の部分をハッシュ形式に変更

ham-cap commented:
早速ご確認いただきありがとうございます🙏
修正いたしました👀
